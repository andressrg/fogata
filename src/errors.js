class ValidationError extends Error {
  constructor(params) {
    super(...arguments);

    if (params instanceof String) {
      this.message = params;
    } else if (params.messages !== undefined) {
      this.messages = params.messages;
    }
  }

  toString() {
    if (this.message) return this.message;
    super.toString();
  }
}

module.exports = { ValidationError };
