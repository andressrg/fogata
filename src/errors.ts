export class ValidationError extends Error {
  messages?: string[] | { [propName: string]: string[] };

  constructor(
    params: string | string[] | { messages: { [propName: string]: string[] } }
  ) {
    super(...arguments);

    if (params instanceof String) {
      this.message = params;
    } else if (params instanceof Array) {
      this.messages = params;
    } else if (params.messages !== undefined) {
      this.messages = params.messages;
    }
  }

  toString() {
    if (this.message) return this.message;
    super.toString();
  }
}
