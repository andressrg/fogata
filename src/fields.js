const { ValidationError } = require('./errors');

class Validator {
  validate() {}
}

class RequiredValidator extends Validator {
  validate(value) {
    super.validate(...arguments);
    if (value === null || value === undefined) {
      throw new ValidationError('Required value.');
    }
  }
}

class Field {
  constructor(params = {}) {
    this.validators = params.validators || [];

    this.required = params.required;
    if (this.required === true) {
      this.validators = [...this.validators, new RequiredValidator()];
    }
  }

  load(input) {
    const errors = this.validators
      .map(validator => {
        try {
          validator.validate(input);
        } catch (err) {
          return err.toString();
        }
      })
      .filter(e => !!e);

    if (errors.length > 0) {
      throw new ValidationError(errors);
    }

    return input;
  }

  dump(input) {
    return input;
  }
}

class StringField extends Field {
  load(input) {
    input = super.load(...arguments);
    return `${input}`;
  }

  dump(input) {
    input = super.dump(...arguments);
    return this.load(input);
  }
}

class DateField extends Field {
  load(input) {
    input = super.load(...arguments);
    return new Date(input);
  }

  dump(input) {
    input = super.dump(...arguments);
    return input.toISOString();
  }
}

class EmailValidator extends Validator {
  validate(value) {
    super.validate(...arguments);

    if (value !== undefined && !value.includes('@')) {
      throw new ValidationError(`${value} is not a valid email address.`);
    }
  }
}

class EmailField extends Field {
  constructor() {
    super();
    this.validators = [...this.validators, new EmailValidator()];
  }
}

class NestedField extends Field {
  constructor(nestedSchema) {
    super(...arguments);
    this.nestedSchema = nestedSchema;
  }

  load(input) {
    input = super.load(...arguments);
    return this.nestedSchema.load(input);
  }

  dump(input) {
    input = super.dump(...arguments);
    return this.nestedSchema.dump(input);
  }
}

module.exports = {
  Field,
  StringField,
  NestedField,
  DateField,
  EmailField
};
