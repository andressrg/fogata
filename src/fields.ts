import { ValidationError } from './errors';
import { Schema } from './index';

export class Validator implements Validator {
  validate(value: any) {}
}

class RequiredValidator extends Validator {
  validate(value: any) {
    super.validate(value);
    if (value === null || value === undefined) {
      throw new ValidationError('Required value.');
    }
  }
}

export interface FieldParams {
  validators?: Array<Validator>;
  required?: boolean;
}

export class Field {
  validators: Validator[];
  required: boolean;

  constructor(params: FieldParams = {}) {
    this.validators = params.validators || [];

    this.required = params.required || false;
    if (this.required === true) {
      this.validators = [...this.validators, new RequiredValidator()];
    }
  }

  load(input: any): any {
    const errors: string[] = this.validators
      .map(validator => {
        try {
          validator.validate(input);
        } catch (err) {
          return err.toString();
        }
      })
      .filter(e => !!e);

    if (errors.length > 0) {
      throw new ValidationError(errors);
    }

    return input;
  }

  dump(input: any): any {
    return input;
  }
}

export class StringField extends Field {
  load(input: any): string {
    input = super.load(input);
    return `${input}`;
  }
}

export class DateField extends Field {
  load(input: any): Date {
    input = super.load(input);
    return new Date(input);
  }

  dump(input: Date): string {
    input = super.dump(input);
    return input.toISOString();
  }
}

export class EmailValidator extends Validator {
  validate(value: string) {
    super.validate(value);

    if (value !== undefined && !value.includes('@')) {
      throw new ValidationError(`${value} is not a valid email address.`);
    }
  }
}

export class EmailField extends Field {
  constructor(params: FieldParams = {}) {
    super(params);
    this.validators = [...this.validators, new EmailValidator()];
  }
}

export class NestedField extends Field {
  nestedSchema: Schema;

  constructor(nestedSchema: Schema) {
    super(...arguments);
    this.nestedSchema = nestedSchema;
  }

  load(input: any): any {
    input = super.load(input);
    return this.nestedSchema.load(input);
  }

  dump(input: any) {
    input = super.dump(input);
    return this.nestedSchema.dump(input);
  }
}
