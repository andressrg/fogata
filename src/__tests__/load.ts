import { Schema } from '../';
import * as fields from '../fields';
import { ValidationError } from '../errors';

class MyAsyncField extends fields.Field {
  async load(input: any) {
    return await Promise.resolve(input);
  }

  async dump(input: any) {
    return await Promise.resolve(input);
  }
}

class ArtistSchema extends Schema {
  static artistName = new fields.StringField();
  static myAsyncAttr = new MyAsyncField();
}

class AlbumSchema extends Schema {
  static get title() {
    return new fields.StringField();
  }

  static get release_date() {
    return new fields.DateField();
  }

  static get artist() {
    return new fields.NestedField(new ArtistSchema());
  }
}

it('loads', async () => {
  expect(
    await new AlbumSchema().load({
      artist: { artistName: 'El Cuarteto De Nos', myAsyncAttr: 'something' },
      title: 'Habla Tu Espejo',
      release_date: '2014-10-06T18:51:17.749Z'
    })
  ).toMatchSnapshot();
});

it('validation throws', async () => {
  class UserSchema extends Schema {
    static email = new fields.EmailField({ required: true });
    static artistName = new fields.StringField({ required: true });
  }

  const resultPromise = new UserSchema().load({
    email: 'foo'
  });
  await expect(resultPromise).rejects.toBeInstanceOf(ValidationError);

  try {
    await new UserSchema().load({
      email: 'foo'
    });
  } catch (err) {
    expect(err).toBeInstanceOf(ValidationError);
    expect(err.messages).toEqual({
      email: ['foo is not a valid email address.'],
      artistName: ['Required value.']
    });
  }

  await expect(new UserSchema().load({})).rejects.toBeInstanceOf(
    ValidationError
  );
});
