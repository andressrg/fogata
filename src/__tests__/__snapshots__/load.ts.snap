// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads 1`] = `
Object {
  "artist": Object {
    "artistName": "El Cuarteto De Nos",
    "myAsyncAttr": "something",
  },
  "release_date": 2014-10-06T18:51:17.749Z,
  "title": "Habla Tu Espejo",
}
`;
